libvast/vast/bitmap_algorithms.hpp

```diff
@@ -292,114 +292,197 @@ select(const Bitmap& bm, typename Bitmap::size_type i) {
-    if (n <= remaining) {
-      i_ += n - 1;
+  // -- explicit range control ------------------------------------------------
+
+  /// Moves the range forward by *k* bits from the current position.
+  /// @param k The number of bits to seek forward.
+  /// @pre `!done() && k > 0`
+  void next(size_type k) {
+    VAST_ASSERT(k > 0);
+    VAST_ASSERT(i_ != npos);
+    VAST_ASSERT(!bits().empty());
+    auto remaining_bits = bits().size() - i_ - 1;
+    if (k <= remaining_bits) {
+      i_ += k - 1;
       next();
-    } else {
-      n -= remaining;
-      i_ = word_type::npos;
-      n_ += rng_.get().size();
-      rng_.next();
-      while (rng_) {
-        if (n > rng_.get().size()) {
-          n -= rng_.get().size();
-        } else {
-          i_ = n - 2;
-          next();
-          break;
-        }
-        n_ += rng_.get().size();
-        rng_.next();
+      return;
+    }
+    for (k -= remaining_bits, i_ = npos, n_ += bits().size(), rng_.next();
+         !rng_.done() && k > 0;
+         k -= bits().size(), n_ += bits().size(), rng_.next()) {
+      if (k <= bits().size()) {
+        i_ = k - 1;
+        return;
       }
     }
   }
 
-  bool done() const {
-    return rng_.done() && i_ == word_type::npos;
-  }
-
-protected:
-  void scan() {
-    while (i_ == word_type::npos) {
-      n_ += rng_.get().size();
+  /// Moves to the range to the next bit of a given value.
+  /// @tparam Bit The bit value to move forward to.
+  template <bool Bit = true>
+  void select() {
+    VAST_ASSERT(i_ != npos);
+    i_ = find_next<Bit>(bits(), i_);
+    while (i_ == npos) {
+      n_ += bits().size();
       rng_.next();
       if (rng_.done())
         return;
-      i_ = find_first(rng_.get());
+      i_ = find_first<Bit>(bits());
+    }
+  }
+
+  /// Moves the range forward by *k* bits having a given value. The effect of
+  /// this function is equivalent to *k* invocations of `next<Bit>()`.
+  /// @tparam Bit The bit value to move forward with.
+  /// @param k The number bits of value *Bit* to move forward.
+  template <bool Bit = true>
+  void select(size_type k) {
+    VAST_ASSERT(k > 0);
+    VAST_ASSERT(i_ != npos);
+    using vast::select;
+    auto prev = rank<Bit>(bits(), i_);
+    auto remaining = rank<Bit>(bits()) - prev;
+    if (k <= remaining) {
+      i_ = select<Bit>(bits(), prev + k);
+      VAST_ASSERT(i_ != npos);
+      return;
+    }
+    for (k -= remaining, i_ = npos, n_ += bits().size(), rng_.next();
+         !rng_.done();
+         n_ += bits().size(), rng_.next()) {
+      VAST_ASSERT(k > 0);
+      if (k > bits().size()) {
+        k -= rank<Bit>(bits());
+      } else {
+        i_ = select<Bit>(bits(), k);
+        if (i_ != npos)
+          break;
+      }
+    }
+  }
+
+  /// Advances the range to a given position and then performs a subsequent
+  /// call to select().
+  /// @param x The ID where to start the call to select().
+  template <bool Bit = true>
+  void select_at(vast::id x) {
```

Comment:
Just for clarification: can I use this function to "randomly seek", i.e., jump back?

Reply:
No, due to the encoding, all bitmaps and corresponding iteration algorithms behave like "forward iterators."